File: src/lib/code/mcs-216/session-1/5/5.c
#include <stdio.h>

void bubble_sort(int arr[], int n) {
    int exchanges = 0, comparisons = 0;
    int outer_iters = 0, inner_iters = 0;

    // Outer loop: Each pass ensures the i-th largest element is in place
    for (int i = 0; i < n; i++) {
        outer_iters++;

        // Inner loop: Reduces range as elements are sorted at the end
        for (int j = 0; j < n - i - 1; j++) {
            inner_iters++;
            comparisons++;

            // Comparison
            if (arr[j] > arr[j + 1]) {
                // Exchange logic using a temporary variable
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                exchanges++;
            }
        }
    }

    printf("Metrics for Bubble Sort:\n");
    printf("(i)   Exchanges: %d\n", exchanges);
    printf("(ii)  Comparisons: %d\n", comparisons);
    printf("(iii) Outer Iterations: %d, Inner Iterations: %d\n", outer_iters, inner_iters);
}

int main() {
    int data[] = {55, 25, 15, 40, 60, 35, 17, 65, 75, 10};
    int size = sizeof(data) / sizeof(data[0]);
    bubble_sort(data, size);
    return 0;
}

File: src/lib/code/mcs-216/session-1/5/5.rs
fn bubble_sort(arr: &mut [i32]) {
    let n = arr.len();
    let (mut exchanges, mut comparisons) = (0, 0);
    let (mut outer_iters, mut inner_iters) = (0, 0);

    // Outer loop: Tracks how many elements are already placed at the end
    for i in 0..n {
        outer_iters += 1;

        // Inner loop: Comparison limit decreases as 'i' increases
        for j in 0..n - i - 1 {
            inner_iters += 1;
            comparisons += 1;

            if arr[j] > arr[j + 1] {
                // Exchange using Rust's safe swap method
                arr.swap(j, j + 1);
                exchanges += 1;
            }
        }
    }

    println!("Exchanges: {}, Comparisons: {}", exchanges, comparisons);
    println!(
        "Outer Loop Iters: {}, Inner Loop Iters: {}",
        outer_iters, inner_iters
    );
}

fn main() {
    let mut nums = [55, 25, 15, 40, 60, 35, 17, 65, 75, 10];
    bubble_sort(&mut nums);
}
